Fixed sync problem with cuda streams

Johngibbard opened the pull request:
Fixed desync problem with streams, also added some kernels and functions im working on to do more computation on GPU, still need to finish attention head calculations.

StickOnAStick committed:
Removed Wrapper.hh with unnecessary CudaFixedVectorMath namespace.

Updated all FixedVectorMath method definitions to be **Inline**.
This was done to prevent linker errors when NVCC creates a seperate
instance of FixedVectorMath, see [LearnCPP Inline](https://www.learncpp.com/cpp-tutorial/inline-functions-and-variables/)
for more information.

Added "USE_CUDA" definition to transformer.cc as a
quick-n-dirty way to distinguish processing strategies.

Updated transformer.cc to utilize cuda when enabled.

StickOnAStick committed:
Removed Wrapper.hh with unnecessary CudaFixedVectorMath namespace.

Updated all FixedVectorMath method definitions to be **Inline**.
This was done to prevent linker errors when NVCC creates a seperate
instance of FixedVectorMath, see [LearnCPP Inline](https://www.learncpp.com/cpp-tutorial/inline-functions-and-variables/)
for more information.

Added "USE_CUDA" definition to transformer.cc as a
quick-n-dirty way to distinguish processing strategies.

Updated transformer.cc to utilize cuda when enabled.

StickOnAStick committed:
Removed unnecessary inline methods due to previous misunderstanding
of inline optimizations within namespaces.

Johngibbard committed:
Fixed sync problem with cuda streams

Johngibbard committed:
Fixed sync problem with cuda streams

Johngibbard committed:
Fixed sync problem with cuda streams

Johngibbard commented on inc/cuda.cu:
```diff
@@ -263,75 +698,78 @@ namespace CudaFixedVectorMath {
+        FixedVector<float> bias) 
+        {
 
-    FixedVector<FixedVector<float>> linear(FixedVector<FixedVector<float>> A, 
```
Linear now uses default Cuda stream for Async functions

Johngibbard commented on inc/cuda.cu:
```diff
@@ -182,23 +619,17 @@ namespace CudaFixedVectorMath {
         dim3 blockDim(16, 16);
         dim3 gridDim((m + blockDim.x - 1) / blockDim.x, (n + blockDim.y - 1) / blockDim.y);  // Grid size based on matrix dimensions
-        sgemm_naive<<<gridDim, blockDim>>>(m, n, k, dA, dB, dOut);
+        sgemm_naive<<<gridDim, blockDim,0,stream1>>>(m, n, k, dA, dB, dOut);
```
Sgemm kernel now uses stream1 instead of default stream

StickOnAStick commented:
Can we get a more clear name for `sgemm`. We know that it's dot-product now, but might have trouble recognizing this later.

StickOnAStick commented:
What version of nvcc did you use to compile this? 

Will not build on my machine.

StickOnAStick commented:
> What version of nvcc did you use to compile this?
> 
> Will not build on my machine.

This branch has a different makefile. 

Specifically, `-L/usr/local/cuda-12.6/lib64` was missing from the `LDLIBS` flags.

#### Fix:

Change `LDLIBS` to:
```make
LDLIBS  += -L$(TRIPLET_DIR)/lib -L$(TRIPLET_DIR)/lib/manual-link -L/usr/local/cuda-12.6/lib64
```

StickOnAStick committed:
Organized Cuda file.

No implementation details changed.

StickOnAStick committed:
Removed Wrapper.hh with unnecessary CudaFixedVectorMath namespace.

Updated all FixedVectorMath method definitions to be **Inline**.
This was done to prevent linker errors when NVCC creates a seperate
instance of FixedVectorMath, see [LearnCPP Inline](https://www.learncpp.com/cpp-tutorial/inline-functions-and-variables/)
for more information.

Added "USE_CUDA" definition to transformer.cc as a
quick-n-dirty way to distinguish processing strategies.

Updated transformer.cc to utilize cuda when enabled.

StickOnAStick committed:
moved definition to correct location

StickOnAStick commented:
Manually merged into main.
